thread resource pools and kernel object memory management

tyrantlizard77 opened the pull request:
See RFC in #6972 

Broadly speaking:
- Introduce concept of thread resource k_mem_pools which can be assigned to threads, and will serve as a kernel-side heap for allocating kernel objects, buffers used by kernel objects, and other transient data structures used behind the scene by system calls
- infrastructure for kernel objects to release allocated resources (or free the object itself) by reference counting; existing permission model used, permission on an object is the same as a reference to an object.
- fixes for k_pipe to add new init function to expose to user space to allocate the pipe's internal buffer, solving #6996 
- fixes for k_msgq, same reasons as k_pipe, solving #6997 
- similar fix for k_stack #7285 
- k_queues now available to user mode doing behind-the-scenes allocations for the kernel-side linked list
- k_object_allocate() now uses thread resource pool

Commit:
mempool: add API for malloc semantics

This works like k_malloc() but allows the user to designate
a specific memory pool to use instead of the kernel heap.

Test coverage provided by existing tests for k_malloc(), which is
now derived from this API.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: internal APIs for thread resource pools

Some kernel APIs may need to allocate memory in order to function
correctly, especially if they are exposed to userspace where
buffers provided by user code cannot be trusted.

Instead of simply drawing from the system heap, specific pools
may instead be assigned to threads, and any requests made on
behalf of the calling thread will draw heap memory from that pool.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: pipes: add k_pipe_alloc_init()

User mode can't be trusted to provide the kernel buffers for
internal use. The syscall for k_pipe_init() has been removed
in favor of a new API to draw the buffer memory from the
calling thread's resource pool.

K_PIPE_DEFINE() now properly locates the allocated buffer into
kernel memory.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add k_msgq_alloc_init()

User mode can't be trusted to provide a memory buffer to
k_msgq_init(). Introduce k_msgq_alloc_init() which allocates
the buffer out of the calling thread's resource pool and expose
that as a system call instead.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

lpereira commented on include/kernel.h:
```diff
@@ -3357,8 +3357,15 @@ struct k_pipe {
+ * @brief Initialize a pipe and allocate a buffer for it
+ *
+ * Storage for the buffer region will be allocated from the calling thread's
+ * resource pool. This memory will not ever be released.
```
I understand why the memory for this object is never released, but this makes me uneasy.  There might be some gotchas (e.g. a k_pipe allocated this way being used by a supervisor thread, and the user thread that allocates this dies), but I would take a look at making these allocations reference-counted.

Maybe if you repurpose the permission bitmap and use __builtin_popcount() as a refcount you could get away with most of the boilerplate?  (If popcount is 0 on thread exit, k_free() can be called.)

andrewboie replied:
I think I see where you are going with this.

> Maybe if you repurpose the permission bitmap 

The problem is that supervisor threads can do anything with any object they want. They do not need to be assigned permission. We could be in the middle of an API call from a supervisor thread and suddenly free the buffer if all we were tracking were permission settings. Maybe we enforce that supervisor mode has to self-grant permissions or undefined behavior may take place? I think I'm OK with that.

> (If popcount is 0 on thread exit, k_free() can be called.)

I see where to do this. We could check it in _thread_perms_all_clear, specifically in the clear_perms_cb(). It would make this a little slower but I don't think that's a big deal.

Thinking some more about this...

andrewboie replied:
Worth noting: currently, this mechanism is orthogonal to user mode.

If we add reference counting / automatic freeing of resources, we would have to either:
- Document that resource releasing only happens if CONFIG_USERSPACE is enabled, but otherwise leave as-is
- #ifdef out all the new APIs which do behind the scenes allocation

I think I prefer the first option.

andrewboie replied:
I can think of situations where this can blow up.

Say we have threads A and B. They mutually have permissions on each other and some k_pipe we'll call P.

A enters an API call to do something with P. A is preempted while in the syscall.
B is scheduled. B calls k_object_access_revoke(P, B) and k_object_access_revoke(P, A)
P no longer has any references, so the final call to k_object_access_revoke() initiates logic to free P's memory buffer.
A swaps back in to continue processing the system call and promptly explodes while in supervisor mode.
Hmmm.

andrewboie replied:
Maybe, to supplement the permission/reference bitmap, we additionally increment/decrement some per-object counter upon API entry/exit. That sounds horrible to maintain.

Could also make k_object_access_revoke() supervisor-only.

Commit:
kernel: remove k_object_access_revoke() as syscall

Forthcoming patches will dual-purpose an object's permission
bitfield as also reference tracking for kernel objects, used to
handle automatic freeing of resources.

We do not want to allow user thread A to revoke thread B's access
to some object O if B is in the middle of an API call using O.

However we do want to allow threads to revoke their own access to
an object, so introduce a new API and syscall for that.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: automatic resource release framework

An object's set of permissions is now also used as a form
of reference counting. If an object's permission bitmap gets
completely cleared, it is now possible to specify object type
specific cleanup functions to be implicitly called.

Currently no objects are enabled yet. Forthcoming patches
will do this on a per object basis.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: get dynamic objs from thread rsrc pools

Dynamic kernel objects no longer is hard-coded to use the kernel
heap. Instead, objects will now be drawn from the calling thread's
resource pool.

Since we now have a reference counting mechanism, if an object
loses all its references and it was dynamically allocated, it will
be automatically freed.

A parallel dlist is added for efficient iteration over the set of
all dynamic objects, allowing deletion during iteration.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
This is still very WIP but has most of the concepts discussed in #6972.

Commit:
list_gen: slist: mark some APIs are private

The new APIs added are used behind the scenes by list_gen.h
to abstract away details on how lists and nodes work.

Name them with z_ prefix as they are not intended for use
by applications.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
sflist: slist-alike that stores flags

This slist variant allows flags to be stored in a node
without taking up any extra space by relying on 4-byte
pointer alignment; flags are stored in the two LSBs.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
This is now ready for review.

andrewboie commented:
Working through some CI issues:

- Thread resource patch is pulling in mempool code that was previously discarded even if usused, need to figure out what's going on, causing a failure on grove light test on Quark D2000
- Weird xtensa crash on a seemingly unrelated test
- native_posix tests failing for some reason
- tests/kernel/mem_protect/mem_protect failing on minnowboard

Commit:
kernel: stacks: add k_stack_alloc() init

Similar to what has been done with pipes and message queues,
user mode can't be trusted to provide a buffer for the kernel
to use. Remove k_stack_init() as a syscall and offer
k_stack_alloc_init() which allocates a buffer from the caller's
resource pool.

Fixes #7285

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross reviewed (approved):
Seems about as clean as we're going to get.  No complaints beyond the aesthetic ones I've made in the RFC.

dbkinder commented on doc/kernel/data_passing/fifos.rst:
```diff
@@ -27,7 +27,9 @@ A fifo must be initialized before it can be used. This sets its queue to empty.
 the first 32 bits of an item for use as a pointer to the next data item in
 the queue. Consequently, a data item that holds N bytes of application data
-requires N+4 bytes of memory.
+requires N+4 bytes of memory. There is no requirement for data items if
```
should we clarify by saying, ``no alignment requirement`` here?

dbkinder commented on doc/kernel/data_passing/lifos.rst:
```diff
@@ -27,7 +27,9 @@ A lifo must be initialized before it can be used. This sets its queue to empty.
 the first 32 bits of an item for use as a pointer to the next data item in
 the queue. Consequently, a data item that holds N bytes of application data
-requires N+4 bytes of memory.
+requires N+4 bytes of memory. There is no requirement for data items if
```
similarly, should we clarify by saying, ``no alignment requirement`` here?

dbkinder commented on doc/kernel/memory/pools.rst:
```diff
@@ -168,6 +173,25 @@ memory block is actually used to satisfy the request.)
+calling thread. For example, some initialization APIs for objects like
+pipes and message queues may need to allocate a private kernel-side buffer,
+or objects like queues may temporarily allocate kernel data structures
+temporarily as items are placed in the queue.
```
one too many uses of ``temporarily`` in this sentence -- I'd delete this last one.

dbkinder commented on doc/kernel/memory/pools.rst:
```diff
@@ -168,6 +173,25 @@ memory block is actually used to satisfy the request.)
+temporarily as items are placed in the queue.
+
+Such memory allocations are drawn from memory pools that are assigned to
+a thread. By default, any thread in the system has no resource pool and
```
This would read better changing ``any thread`` to ``a thread``

dbkinder reviewed (dismissed):
Looks good -- a few suggestions.

andrewboie commented:
@dbkinder  updated

lpereira commented on include/kernel.h:
```diff
@@ -2863,6 +3031,7 @@ struct k_msgq {
 	char *read_ptr;
 	char *write_ptr;
 	u32_t used_msgs;
+	int flags;
```
Not an issue with just one flag, but it's often a good idea to use unsigned integers for bitmasks.

lpereira commented on kernel/pipes.c:
```diff
@@ -127,31 +128,63 @@ SYS_INIT(init_pipes_module, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
 }
 #endif
 
+void k_pipe_cleanup(struct k_pipe *pipe)
```
No plumbing patch is complete without a pipe cleanup function.

lpereira commented on kernel/pipes.c:
```diff
@@ -127,31 +128,63 @@ SYS_INIT(init_pipes_module, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
+	__ASSERT_NO_MSG(sys_dlist_is_empty(&pipe->wait_q.readers));
+	__ASSERT_NO_MSG(sys_dlist_is_empty(&pipe->wait_q.writers));
+
+	if (pipe->flags & K_PIPE_FLAG_ALLOC) {
```
I wonder if we can write a function that returns true if a pointer is part of a mempool... then we won't need these flags.

andrewboie replied:
I can't think of a robust way of doing this unfortunately.

lpereira reviewed (dismissed):
I still have to sleep on the approach/added APIs, but I have some comments meanwhile

andrewboie commented on kernel/queue.c:
```diff
@@ -102,29 +154,80 @@ void k_queue_insert(struct k_queue *queue, void *prev, void *data)
+
+void k_queue_insert(struct k_queue *queue, void *prev, void *data)
+{
+	queue_insert(queue, prev, data, false);
```
Yes because there's nothing I can do with this as the API itself is void.
It doesn't matter anyway since we pass alloc=false, and we only return -ENOMEM if an allocation is attempted.

lpereira replied:
Yeah. We'll have to revisit these APIs returning void in the future. Failing silently and not giving an opportunity for code to handle that is not going to fly in functional safe scenarios

andrewboie commented on kernel/userspace.c:
```diff
@@ -81,6 +92,8 @@ static size_t obj_size_get(enum k_objects otype)
 		return sizeof(struct k_mutex);
 	case K_OBJ_PIPE:
 		return sizeof(struct k_pipe);
+	case K_OBJ_QUEUE:
```
OK going to look into this.

ioannisg commented:
CC  @pizi-nordic

agross-oss reviewed (commented):
What are the implications of stack alignment requirements for the dynamic allocations?

Commit:
userspace: generate list of kernel object sizes

This used to be done by hand but can easily be generated like
we do other switch statements based on object type.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
The CI failure is baffling me.
It is not SDK dependent. The ARC version of objdump is segfaulting when generating zephyr.lst.
From bisection the offending patch is the one that introduces k_queue support.
Really need to get this in, so I'm going to drop the k_queue patch from this PR and introduce separately while I figure this out.
